aws-sigv4/src/lib.rs

```diff
@@ -123,28 +124,34 @@ where
         date,
         settings,
     } = config;
-    // Step 1: https://docs.aws.amazon.com/en_pv/general/latest/gr/sigv4-create-canonical-request.html.
-    let creq = CanonicalRequest::from(req, &settings).unwrap();
+    let date = DateTime::<Utc>::from(*date);
+    let (creq, extra_headers) = CanonicalRequest::from(req, body, settings, date, *security_token)?;
 
     // Step 2: https://docs.aws.amazon.com/en_pv/general/latest/gr/sigv4-create-string-to-sign.html.
     let encoded_creq = &encode_with_hex(creq.fmt());
-    let date = DateTime::<Utc>::from(date);
     let sts = StringToSign::new(date, region, svc, encoded_creq);
 
     // Step 3: https://docs.aws.amazon.com/en_pv/general/latest/gr/sigv4-calculate-signature.html
     let signing_key = generate_signing_key(secret_key, date.date(), region, svc);
     let signature = calculate_signature(signing_key, &sts.fmt().as_bytes());
 
     // Step 4: https://docs.aws.amazon.com/en_pv/general/latest/gr/sigv4-add-signature-to-request.html
-    let authorization = build_authorization_header(access_key, creq, sts, &signature);
-    let x_azn_date = date.fmt_aws();
-
-    let mut tok = security_token.map(|it| it.to_string());
-    iter::once((SignatureKey::Authorization, authorization))
-        .chain(iter::once((SignatureKey::AmzDate, x_azn_date)))
+    let mut authorization: HeaderValue =
+        build_authorization_header(access_key, &creq, sts, &signature).parse()?;
+    authorization.set_sensitive(true);
+
+    let date = (X_AMZ_DATE, extra_headers.x_amz_date);
+    let mut security_token = extra_headers
+        .x_amz_security_token
+        .map(|tok| (X_AMZ_SECURITY_TOKEN, tok));
+    let mut content = extra_headers
+        .x_amz_content_256
+        .map(|content| (X_AMZ_CONTENT_SHA_256, content));
+    Ok(iter::once(("authorization", authorization))
```

Comment:
I should have caught this sooner, but can we assign the result of `iter::once` to a `let` binding and then return the bound value in the `Ok()`? More of a style thing.